### PR TITLE
settings elements migration

### PR DESCRIPTION
**Pull request by nfarabullini**
- moved the following elements from `model/common/src/icon4py/model/common/settings.py`to `tools/src/icon4pytools/py2fgen/wrappers/settings.py`:
 ```
config = Icon4PyConfig()
backend = config.gt4py_runner
dace_orchestration = config.icon4py_dace_orchestration
xp = config.array_ns
device = config.device
limited_area = config.limited_area
```
- subsequent edit of imports and related in the code
- cleanup of left-over stencils backend specification on program definition

### CODE REVIEW

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

File: model/atmosphere/diffusion/tests/diffusion_tests/mpi_tests/test_parallel_diffusion.py [hunk 1]
```diff
@@ -168,7 +169,7 @@ def test_parallel_diffusion_multiple_steps(
     caplog,
     backend,
 ):
-    if settings.dace_orchestration is None:
+    if dace_orchestration is None:
```
Comment (nfarabullini) [inline model/atmosphere/diffusion/tests/diffusion_tests/mpi_tests/test_parallel_diffusion.py:172]:
```suggestion
    if "dace" not in backend.name.lower():
```

File: model/atmosphere/diffusion/tests/diffusion_tests/test_diffusion.py [hunk 2]
```diff
@@ -511,9 +511,9 @@ def test_run_diffusion_multiple_steps(
     backend,
     icon_grid,
 ):
-    if settings.dace_orchestration is None:
-        raise pytest.skip("This test is only executed for `--dace-orchestration=True`.")
-
+    dace_orchestration = False
+    if not dace_orchestration:
```
Comment (nfarabullini) [inline model/atmosphere/diffusion/tests/diffusion_tests/test_diffusion.py:515]:
```suggestion
    if "dace" not in backend.name.lower():
```

File: model/common/src/icon4py/model/common/settings.py [hunk 3]
```diff
@@ -6,13 +6,7 @@
 # Please, refer to the LICENSE file in the root directory.
 # SPDX-License-Identifier: BSD-3-Clause
 
-from icon4py.model.common.config import Icon4PyConfig
+import numpy as xp
```
Comment (nfarabullini) [inline model/common/src/icon4py/model/common/settings.py]:
```suggestion
import numpy as np
```

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

File: model/common/src/icon4py/model/common/grid/base.py [hunk 4]
```diff
@@ -137,7 +141,7 @@ def _get_offset_provider(self, dim, from_dim, to_dim):
             dim, self.connectivities[dim].dtype
         )
         return gtx.NeighborTableOffsetProvider(
-            xp.asarray(self.connectivities[dim]),
+            np.asarray(self.connectivities[dim]),
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/base.py:144]: Can you remove the `asarray` here all together? Does that work? We should allocate the connectivity arrays already on the right device such that the do not need to be transferred here.

File: model/common/src/icon4py/model/common/grid/geometry.py [hunk 5]
```diff
@@ -8,6 +8,7 @@
 import functools
 from typing import Any, Callable, Literal, Mapping, Optional, Sequence, TypeAlias, TypeVar
 
+import numpy as np
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/geometry.py:11]: Can you add a TODO here that we check whether this works on GPU? You an put my name.
```suggestion
# TODO (@halungge ) test on GPU (NEP 18 ?) 
import numpy as np
```

File: model/common/src/icon4py/model/common/grid/refinement.py [hunk 6]
```diff
@@ -79,21 +79,21 @@ def is_ordered(self) -> bool:
         return self.value not in _UNORDERED[self.dim]
 
 
-def is_unordered_field(field: xp.ndarray, dim: dims.Dimension) -> xp.ndarray:
-    assert field.dtype in (xp.int32, xp.int64), f"not an integer type {field.dtype}"
-    return xp.where(
-        field == _UNORDERED[dim][0], True, xp.where(field == _UNORDERED[dim][1], True, False)
+def is_unordered_field(field: np.ndarray, dim: dims.Dimension) -> np.ndarray:
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/refinement.py:82]:
```suggestion
def is_unordered_field(field: NDArray, dim: dims.Dimension) -> NDArray:
```

File: model/common/src/icon4py/model/common/grid/refinement.py [hunk 7]
```diff
@@ -79,21 +79,21 @@ def is_ordered(self) -> bool:
...
+    assert field.dtype in (np.int32, np.int64), f"not an integer type {field.dtype}"
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/refinement.py:83]:
```suggestion
    assert field.dtype in (gtx.int32, gtx.int64), f"not an integer type {field.dtype}"
```

File: model/common/src/icon4py/model/common/grid/refinement.py [hunk 8]
```diff
@@ -79,21 +79,21 @@ def is_ordered(self) -> bool:
...
-def convert_to_unnested_refinement_values(field: xp.ndarray, dim: dims.Dimension) -> xp.ndarray:
+def convert_to_unnested_refinement_values(field: np.ndarray, dim: dims.Dimension) -> np.ndarray:
     """Convenience function that converts the grid refinement value from a coarser
     parent grid to the canonical values used in an unnested setup.
 
     The nested values are used for example in the radiation grids.
     """
-    assert field.dtype in (xp.int32, xp.int64), f"not an integer type {field.dtype}"
-    return xp.where(field == _UNORDERED[dim][1], 0, xp.where(field < 0, -field, field))
+    assert field.dtype in (np.int32, np.int64), f"not an integer type {field.dtype}"
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/refinement.py:95]:
```suggestion
    assert field.dtype in (gtx.int32, gtx.int64), f"not an integer type {field.dtype}"
```

File: model/common/src/icon4py/model/common/grid/refinement.py [hunk 9]
```diff
@@ -79,21 +79,21 @@ def is_ordered(self) -> bool:
...
-        field == _UNORDERED[dim][0], True, xp.where(field == _UNORDERED[dim][1], True, False)
+def is_unordered_field(field: np.ndarray, dim: dims.Dimension) -> np.ndarray:
+    assert field.dtype in (np.int32, np.int64), f"not an integer type {field.dtype}"
+    return np.where(
+        field == _UNORDERED[dim][0], True, np.where(field == _UNORDERED[dim][1], True, False)
     )
 
 
-def convert_to_unnested_refinement_values(field: xp.ndarray, dim: dims.Dimension) -> xp.ndarray:
+def convert_to_unnested_refinement_values(field: np.ndarray, dim: dims.Dimension) -> np.ndarray:
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/refinement.py:89]:
```suggestion
def convert_to_unnested_refinement_values(field: NDArray, dim: dims.Dimension) -> NDArray:
```

File: model/common/src/icon4py/model/common/grid/vertical.py [hunk 10]
```diff
@@ -246,30 +246,30 @@ def size(self, dim: gtx.Dimension) -> int:
 
     @classmethod
     def _determine_start_level_of_moist_physics(
-        cls, vct_a: xp.ndarray, top_moist_threshold: float, nshift_total: int = 0
+        cls, vct_a: np.ndarray, top_moist_threshold: float, nshift_total: int = 0
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/vertical.py:249]:
```suggestion
        cls, vct_a: NDArray, top_moist_threshold: float, nshift_total: int = 0
```

File: model/common/src/icon4py/model/common/grid/vertical.py [hunk 11]
```diff
@@ -246,30 +246,30 @@ def size(self, dim: gtx.Dimension) -> int:
...
     ) -> gtx.int32:
         n_levels = vct_a.shape[0]
         interface_height = 0.5 * (vct_a[: n_levels - 1 - nshift_total] + vct_a[1 + nshift_total :])
-        return gtx.int32(xp.min(xp.where(interface_height < top_moist_threshold)[0]).item())
+        return gtx.int32(np.min(np.where(interface_height < top_moist_threshold)[0]).item())
 
     @classmethod
-    def _determine_damping_height_index(cls, vct_a: xp.ndarray, damping_height: float) -> gtx.int32:
+    def _determine_damping_height_index(cls, vct_a: np.ndarray, damping_height: float) -> gtx.int32:
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/vertical.py:256]:
```suggestion
    def _determine_damping_height_index(cls, vct_a: NDArray, damping_height: float) -> gtx.int32:
```

File: model/common/src/icon4py/model/common/grid/vertical.py [hunk 12]
```diff
@@ -246,30 +246,30 @@ def size(self, dim: gtx.Dimension) -> int:
...
             0
             if damping_height > vct_a[0]
-            else gtx.int32(xp.argmax(xp.where(vct_a >= damping_height)[0]).item())
+            else gtx.int32(np.argmax(np.where(vct_a >= damping_height)[0]).item())
         )
 
     @classmethod
     def _determine_end_index_of_flat_layers(
-        cls, vct_a: xp.ndarray, flat_height: float
+        cls, vct_a: np.ndarray, flat_height: float
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/vertical.py:266]:
```suggestion
        cls, vct_a: NDArray, flat_height: float
```

File: model/common/src/icon4py/model/common/test_utils/helpers.py [hunk 13]
```diff
@@ -86,30 +85,35 @@ def random_field(
...
 
 def zero_field(
     grid: BaseGrid,
     *dims: gt_common.Dimension,
     dtype=wpfloat,
     extend: Optional[dict[gt_common.Dimension, int]] = None,
+    backend=None,
 ) -> gt_common.Field:
-    return as_field(dims, xp.zeros(shape=_shape(grid, *dims, extend=extend), dtype=dtype))
+    return as_field(
```
Comment (halungge) [inline model/common/src/icon4py/model/common/test_utils/helpers.py:105]:
```suggestion
    domain = {d: _shape(grid, dims, extend) for d in dims}
    return gtx.zeros(domain, dtype=dtype, allocator=backend)
```
should work.

**Review by halungge — CHANGES_REQUESTED** [review]
Tests look all good. A couple of fixes, mostly.


- In `tools.py2fgen` use the `xp` everywhere as before just the location where you import it from has changed, it does not longer import from  the xp icon4py.model.common` but from its own `icon4pytools.py2fgen.settings`
- where appropriate use a  NDArray = Union[cupy.ndarray, numpy.ndarray] in the type annotations.

File: model/atmosphere/diffusion/tests/diffusion_tests/mpi_tests/test_parallel_diffusion.py [hunk 14]
```diff
@@ -168,7 +168,7 @@ def test_parallel_diffusion_multiple_steps(
     caplog,
     backend,
 ):
-    if settings.dace_orchestration is None:
+    if "dace" not in backend.name.lower():
         raise pytest.skip("This test is only executed for `--dace-orchestration=True`.")
```
Comment (kotsaloscv) [inline model/atmosphere/diffusion/tests/diffusion_tests/mpi_tests/test_parallel_diffusion.py:172]: Could you please remove this option from the pytest configuration?
This is not needed anymore.
Thank you!

File: model/common/src/icon4py/model/common/test_utils/pytest_config.py [hunk 15]
```diff
@@ -127,8 +125,8 @@ def pytest_addoption(parser):
     try:
         parser.addoption(
```
Comment (kotsaloscv) [inline model/common/src/icon4py/model/common/test_utils/pytest_config.py:126]: This is not needed anymore, i.e. `--dace-orchestration `

**Review by kotsaloscv — CHANGES_REQUESTED** [review]
LGTM

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Review by kotsaloscv — APPROVED** [review]

File: model/atmosphere/dycore/src/icon4py/model/atmosphere/dycore/velocity_advection.py [hunk 16]
```diff
@@ -460,7 +460,7 @@ def run_predictor_step(
 
     def _update_levmask_from_cfl_clipping(self):
         self.levmask = gtx.as_field(
-            domain=(dims.KDim,), data=(xp.any(self.cfl_clipping.ndarray, 0)), dtype=bool
+            domain=(dims.KDim,), data=(np.any(self.cfl_clipping.ndarray, 0)), dtype=bool
```
[the remark this answers is not in the document]
Comment (nfarabullini) [reply model/atmosphere/dycore/src/icon4py/model/atmosphere/dycore/velocity_advection.py:463]: `cfl_clipping` has dimensions `[CellDim, KDim]` but we cannot return a KDim only field in gt4py

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

File: tools/src/icon4pytools/py2fgen/wrappers/common.py [hunk 17]
```diff
@@ -10,35 +10,39 @@
...
-    c2e: xp.ndarray,
-    e2c: xp.ndarray,
-    c2e2c: xp.ndarray,
-    e2c2e: xp.ndarray,
-    e2v: xp.ndarray,
-    v2e: xp.ndarray,
-    v2c: xp.ndarray,
-    e2c2v: xp.ndarray,
-    c2v: xp.ndarray,
+    cell_starts: NDArray,
```
Comment (halungge) [inline tools/src/icon4pytools/py2fgen/wrappers/common.py:31]: you have the `xp` import as before in `wrappers.common`: 
You can use `xp.ndarray` as before, then you dont need to import another module (gt4py_field_allocation.py) from the model here, which is preferable.

File: tools/src/icon4pytools/py2fgen/wrappers/common.py [hunk 18]
```diff
@@ -78,8 +82,8 @@ def construct_icon_grid(
     e2c2e = adjust_fortran_indices(e2c2e, offset)
 
     # stacked arrays
-    c2e2c0 = xp.column_stack((xp.asarray(range(c2e2c.shape[0])), c2e2c))
-    e2c2e0 = xp.column_stack((xp.asarray(range(e2c2e.shape[0])), e2c2e))
+    c2e2c0 = xp.column_stack((xp.asarray(range(c2e2c.shape[0])), c2e2c.ndarray))
```
Comment (halungge) [inline tools/src/icon4pytools/py2fgen/wrappers/common.py:85]: the connectvities that are used here are `xp.ndarray` not fields.You don need the extra .ndarray.
```suggestion
    c2e2c0 = xp.column_stack((xp.asarray(range(c2e2c.shape[0])), c2e2c))
```

File: tools/src/icon4pytools/py2fgen/wrappers/common.py [hunk 19]
```diff
@@ -132,12 +136,12 @@ def construct_icon_grid(
 
 
 def construct_decomposition(
-    c_glb_index: xp.ndarray,
-    e_glb_index: xp.ndarray,
-    v_glb_index: xp.ndarray,
-    c_owner_mask: xp.ndarray,
-    e_owner_mask: xp.ndarray,
-    v_owner_mask: xp.ndarray,
+    c_glb_index: NDArray,
```
Comment (halungge) [inline tools/src/icon4pytools/py2fgen/wrappers/common.py:139]:
```suggestion
    c_glb_index: xp.ndarray,
```
Same here you have the xp import from settings here.

File: tools/src/icon4pytools/py2fgen/wrappers/dycore_wrapper.py [hunk 20]
```diff
@@ -57,15 +57,16 @@
...
 from icon4py.model.common.grid.icon import GlobalGridParams
 from icon4py.model.common.grid.vertical import VerticalGrid, VerticalGridConfig
-from icon4py.model.common.settings import backend
 from icon4py.model.common.states.prognostic_state import PrognosticState
 from icon4py.model.common.test_utils.helpers import (
     as_1D_sparse_field,
     flatten_first_two_dims,
     zero_field,
 )
+from icon4py.model.common.utils import PredictorCorrectorPair, TimeStepPair
```
Comment (halungge) [inline tools/src/icon4pytools/py2fgen/wrappers/dycore_wrapper.py:66]:
```suggestion
from icon4py.model.common import utils import as common_utils
```

File: tools/src/icon4pytools/py2fgen/wrappers/dycore_wrapper.py [hunk 21]
```diff
@@ -403,8 +404,8 @@ def solve_nh_run(
         mass_fl_e=mass_fl_e,
         ddt_vn_phy=ddt_vn_phy,
         grf_tend_vn=grf_tend_vn,
-        ddt_vn_apc_pc=common_utils.PredictorCorrectorPair(ddt_vn_apc_ntl1, ddt_vn_apc_ntl2),
-        ddt_w_adv_pc=common_utils.PredictorCorrectorPair(ddt_w_adv_ntl1, ddt_w_adv_ntl2),
+        ddt_vn_apc_pc=PredictorCorrectorPair(ddt_vn_apc_ntl1, ddt_vn_apc_ntl2),
+        ddt_w_adv_pc=PredictorCorrectorPair(ddt_w_adv_ntl1, ddt_w_adv_ntl2),
```
Comment (halungge) [inline tools/src/icon4pytools/py2fgen/wrappers/dycore_wrapper.py:408]:
```suggestion
        ddt_vn_apc_pc=common_utils.PredictorCorrectorPair(ddt_vn_apc_ntl1, ddt_vn_apc_ntl2),
        ddt_w_adv_pc=common_utils.PredictorCorrectorPair(ddt_w_adv_ntl1, ddt_w_adv_ntl2),
```

File: tools/src/icon4pytools/py2fgen/wrappers/dycore_wrapper.py [hunk 22]
```diff
@@ -428,7 +429,7 @@ def solve_nh_run(
         rho=rho_new,
         exner=exner_new,
     )
-    prognostic_states = common_utils.TimeStepPair(prognostic_state_nnow, prognostic_state_nnew)
+    prognostic_states = TimeStepPair(prognostic_state_nnow, prognostic_state_nnew)
```
Comment (halungge) [inline tools/src/icon4pytools/py2fgen/wrappers/dycore_wrapper.py:432]:
```suggestion
    prognostic_states = common_utils.TimeStepPair(prognostic_state_nnow, prognostic_state_nnew)
```

**Review by halungge — COMMENTED** [review]
Some small things still:
- Can you fix the import for `NDArray` everywhere: don't import the TypeAlias directly but import the module? 
- In tools you don't need the `NDArray`: there we do the xp import as before so we can use xp.ndarray as type annotations.
There are some places where a `isinstance` check needs to explicitly check for numpy.ndarray.

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

File: model/common/src/icon4py/model/common/grid/base.py [hunk 23]
```diff
@@ -114,7 +113,12 @@ def offset_providers(self):
 
     @utils.chainable
     def with_connectivities(self, connectivity: Dict[gtx.Dimension, np.ndarray]):
-        self.connectivities.update({d: k.astype(gtx.int32) for d, k in connectivity.items()})
+        self.connectivities.update(
+            {
+                d: k.ndarray.astype(gtx.int32) if hasattr(k, "ndarray") else k.astype(gtx.int32)
```
[the remark this answers is not in the document]
Comment (halungge) [reply model/common/src/icon4py/model/common/grid/base.py:119]: Can you still remove and rerun without the `hasattr`? 
```suggestion
               self.connectivities.update({d: k.astype(gtx.int32) for d, k in connectivity.items()})
```

**Comment by nfarabullini** [comment]
launch jenkins spack

**Comment by nfarabullini** [comment]
cscs-ci run default

File: model/common/src/icon4py/model/common/grid/icon.py [hunk 24]
```diff
@@ -127,11 +127,22 @@ def __eq__(self, other: "IconGrid"):
 
     @utils.chainable
     def with_start_end_indices(
-        self, dim: gtx.Dimension, start_indices: np.ndarray, end_indices: np.ndarray
+        self,
+        dim: gtx.Dimension,
+        start_indices: field_alloc.NDArray,
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/icon.py:132]: According to Chia Rui those must be on CPU, you could keep a `np.ndarray` here.

File: model/common/src/icon4py/model/common/grid/icon.py [hunk 25]
```diff
@@ -127,11 +127,22 @@ def __eq__(self, other: "IconGrid"):
...
+        end_indices: field_alloc.NDArray,
     ):
         log.debug(f"Using start_indices {dim} {start_indices}, end_indices {dim} {end_indices}")
-        self._start_indices[dim] = start_indices.astype(gtx.int32)
-        self._end_indices[dim] = end_indices.astype(gtx.int32)
+        self._start_indices[dim] = (
+            start_indices.ndarray.astype(gtx.int32)
+            if hasattr(start_indices, "ndarray")
+            else start_indices.astype(gtx.int32)
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/icon.py:139]: Same as above: Can you drop the if branch and rerun. 

```suggestion
start_indices.astype(gtx.int32)
```

File: model/common/src/icon4py/model/common/grid/icon.py [hunk 26]
```diff
@@ -127,11 +127,22 @@ def __eq__(self, other: "IconGrid"):
...
+        )
+        self._end_indices[dim] = (
+            end_indices.ndarray.astype(gtx.int32)
+            if hasattr(end_indices, "ndarray")
+            else end_indices.astype(gtx.int32)
```
Comment (halungge) [inline model/common/src/icon4py/model/common/grid/icon.py:144]:
```suggestion
            end_indices.astype(gtx.int32)
```

**Review by halungge — COMMENTED** [review]
Can you rerun without those `hasattr` calls in `grid/icon.py` and `grid/base.py` Otherwise I think it is good to go, the rest we can fine tune in another PR.

**Comment by nfarabullini** [comment]
cscs-ci run default

**Comment by halungge** [comment]
cscs-ci run default

**Comment by nfarabullini** [comment]
launch jenkins spack

**Review by halungge — APPROVED** [review]

**Pull request merged**